### ISSUE TITLE
Use uniforms for forms

### ISSUE THREAD

**Issue by radekmie** [VulcanJS/Vulcan#1603]
Hey there!

I don't know if you heard about [uniforms](https://github.com/vazco/uniforms) - a package for managing forms in React. I'm the author of this package and _(I don't know if I should say that, but I heard that few times)_ that it's becoming a standard in the Meteor community _(it's not a Meteor package, but it's the only one I know, that work seamlessly with SimpleSchema)_.

Why? I've looked deeply into `vulcan:forms` and I've noticed three things: one, a lot of logic have to be maintained within this package, two, it's already battle tested in few really big applications, three, uniforms have already 5 themes: AntD, Bootstrap 3, Bootstrap 4, Material UI, Semantic UI and plan HTML one. Also, a lot of code can be easily removed because for example generating forms from schema _(both GraphQL and SimpleSchema)_ is already implemented.

Downsides? API will _(at least a bit)_ change. We can expose similar _(or even the same)_ API, but it will require some work.

Thoughts?

**Comment by SachaG** [comment]
I think that could be a great idea! I would need to take some time to take a better look at the library to decide though, maybe we could get a discussion started in the Slack chatroom? (http://slack.telescopeapp.org)

To be honest, there's three main issues that would prevent us from switching:

1. It's not a good match from a features set perspective (but I don't think that's the case)
2. We don't want to depend on a third-party package in case it stops being maintained (but I also don't think that's the case ;)
3. We don't want to depend on a Meteor package in case we one day want to move out of Meteor. 

So basically, the main issue is 3. The overall direction for Vulcan is to move more towards the "standard" React ecosystem (Redux, Apollo, etc.) so we've been generally trying to avoid depending on any Meteor-specific packages whenever possible.

**Comment by SachaG** [comment]
Oh wait it looks like I'm wrong and Uniforms is already independent from Meteor? If that's the case then I don't really have much objections :)

**Comment by radekmie** [comment]
Yep, it's a npm package. The only Meteor connection is the detection of SimpleSchema package.

**Comment by iamrommel** [comment]
I've been using redux-form (http://redux-form.com/6.6.1/) on several projects i have and i find it cool. I will try do some implementation using vulcan or even publish it as a package if i find it worth it.

**Comment by radekmie** [comment]
You mean uniforms with redux?

**Comment by SachaG** [comment]
@radekmie yes. Would that work?

**Comment by radekmie** [comment]
@SachaG Like a charm - simply use forms `model` as a value and `onChange` as a... `onChange`.

Every uniforms form can be treated as an input. Even more - it's more performant to keep the state on your own instead of using `AutoForm`.

**Comment by SachaG** [comment]
Another question – would it be easy to integrate Uniforms with Meteor 1.5's conditional loading? For example, if one of your form fields is a datepicker, it'd be nice to conditionally load the datepicker library?

**Comment by radekmie** [comment]
You mean the date picker from uniforms or a custom one? You can import only some parts of these components, but some of them, like `AutoField`, which may render any field, will import all of them. If you mean a custom one, then no, it's not a problem

**Comment by dchan14** [comment]
so is it possible to use uniforms? if so is there an example?  greatly appreciated!

**Comment by dchan14** [comment]
uniforms requires a schema and i didn't know how to pass it correctly so i thought this was a lost cause. i like smartforms it does everything i want but prefill props of an edit form. and i am just assuming if i could cherry pic fields using uniforms it would just work ... thanks in advance!

**Comment by radekmie** [comment]
@dchan14 it's not integrated yet, but you can still use uniforms. To get schema from collection, simply call `collection.simpleSchema()` and pass it as `schema` prop to a uniforms form.

**Comment by SachaG** [comment]
Honestly I just haven't had time to really look into Uniforms, so for now I feel more comfortable keeping SmartForms since I'm already familiar with the code, and can be sure it only does what's needed. If someone wants to try out integrating Uniforms and then compiling a list of pros/cons that'd be super helpful though.

**Comment by darwis059** [comment]
@radekmie I try to pass collection.simpleSchema(), but i got this warning on browser console. 
`Warning: Unknown props `viewableBy`, `insertableBy`, `editableBy` on <input> `

**Comment by radekmie** [comment]
@darwis059 That's because Vulcan has some additional schema fields. To get rid of them in your fields, do this in your project:

```js
import filterDOMProps from 'uniforms/filterDOMProps';

filterDOMProps.register(
    'editableBy',
    'insertableBy',
    'viewableBy'
);
```

**Comment by dchan14** [comment]
i got to admit uniforms is cool looking but it's hard to give up the permissions aspect of vulcan on my end and replacing the render function extending smart forms appear to make more sense now.

**Comment by eric-burel** [comment]
Closing for staleness. Improving the SmartForm is definitely a good idea but we first would need to clean it up, split it's features, and write tons of tests. Otherwise we may be subjects to massive regressions while the SmartForm is used a LOT by all Vulcan applications.
